test: suppress pytest warnings

DaniBodor opened the pull request:
This PR avoids getting the 100s of warnings that are thrown when we run pytest.

- some test functions were basically used as the callback function of a "test all envs" function. I made those internal only functions, to avoid return statement warnings (this would become errors in the next pytest update).
- I suppressed some common warnings as a temporary fix to avoid the clutter of warnings when running pytest, but these should be dealt with more carefully. These seem to be incorrect or deprecated use of the gymnasium objects (potentially differences to gym, i haven't looked into it in enough detail).
- previous test_speed was not an actual test, but more of a way to log how long things take. Also, it was quite slow. So I renamed it, thereby avoiding it being run each time pytest is run.

NOT SOLVED: A number of the tests (especially in `test_env`) catch all errors in the middle of the test function. This means that the test cannot fail and is basically not testing anything (instead in `print`s an error message, but pytest doesn't show that). I tried not catching errors, but then the test does fail and I didn't look into it any further (yet). I added FIXME statements to those.

Commit:
tests: suppress common warnings

Commit:
deal with specific warnings

Commit:
log speed not a python test

Commit:
test_envs

Commit:
test_envs remove warnings

Commit:
minor improvements to test_env

gcroci2 commented on tests/test_envs.py:
```diff
@@ -1,32 +1,29 @@
-        env = ngym.all_envs()[0]
     if isinstance(env, str):
-        env = make_env(env, **kwargs)
+        env = ngym.make(env)
```
Why `**kwargs` has been removed?

DaniBodor replied:
because it is empty. No actual kwargs are sent to the function in most cases. The `make_env` function, was an internal only function that literally called `ngym.make` and did nothing else, so I removed that function and only sent kwargs in case there were any to send.

gcroci2 commented on tests/test_envs.py:
```diff
@@ -52,127 +49,148 @@ def test_run(env=None, num_steps=100, verbose=False, **kwargs):
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", message=".*get variables from other wrappers is deprecated*")
+        warnings.filterwarnings("ignore", message=".*The environment creator metadata doesn't include `render_modes`*")
+        for env_name in ENVS:
```
Why `sorted()` has been removed?

DaniBodor replied:
This function tests each environment (from the ENVS list). The order in which they are tested did not seem relevant to me. Also, it seems that the way `ENV` is constructed already sorts the entries.

gcroci2 commented on tests/test_envs.py:
```diff
@@ -52,127 +49,148 @@ def test_run(env=None, num_steps=100, verbose=False, **kwargs):
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", message=".*get variables from other wrappers is deprecated*")
+        warnings.filterwarnings("ignore", message=".*The environment creator metadata doesn't include `render_modes`*")
+        for env_name in ENVS:
```
I'd leave the sorting here, it's nice to print the in alphabetical order

DaniBodor replied:
`ENVS` is already a sorted list. Also, the printing doesn't actually happen when running pytest, only when running the function in a different way. (`ENVS == sorted(ENVS)` returns `True`)

gcroci2 commented on tests/test_envs.py:
```diff
@@ -52,127 +49,148 @@ def test_run(env=None, num_steps=100, verbose=False, **kwargs):
     """Main function for testing if an environment is healthy."""
-    if env is None:
-        env = ngym.all_envs()[0]
+    with warnings.catch_warnings():
```
Since this function is called by `test_dataset_all`, which already catches warnings, I don't think the catch is needed here, right?

DaniBodor replied:
I guess these particular warning messages weren't caught there. I will adjust this so that all warnings catches are consolidated.

gcroci2 commented on tests/test_envs.py:
```diff
@@ -52,127 +49,148 @@ def test_run(env=None, num_steps=100, verbose=False, **kwargs):
         print(f"Running env: {env_name}")
-        obs1, rews1, acts1 = test_seeding(env_name, seed=0)
-        obs2, rews2, acts2 = test_seeding(env_name, seed=0)
+        obs1, rews1, acts1 = _test_seeding(env_name)
```
I'd catch the warning here, in the upper function (`test_seeding_all`) instead of in `_test_seeding`, as it's done already for `test_run_all` and `test_trialenv_all`

gcroci2 reviewed (approved):
Great, now running the tests will be much lighter :D

> * some test functions were basically used as the callback function of a "test all envs" function. I made those internal only functions, to avoid return statement warnings (this would become errors in the next pytest update).

Why they would become errors in the next pytest update?

> * previous test_speed was not an actual test, but more of a way to log how long things take. Also, it was quite slow. So I renamed it, thereby avoiding it being run each time pytest is run.

As it is now, we won't notice any error when changing the code, and we will need to remember to run it every time to keep it up-to-date with the code-base. I'd either remove it or leave it as a test (naming it back to test_speed).
> 
> NOT SOLVED: A number of the tests (especially in `test_env`) catch all errors in the middle of the test function. This means that the test cannot fail and is basically not testing anything (instead in `print`s an error message, but pytest doesn't show that). I tried not catching errors, but then the test does fail and I didn't look into it any further (yet). I added FIXME statements to those.

Useful that you have already flagged it. In general, I'd avoid looking too much into such issues for now, since the priority at the moment is to let the package work for our experiments, and edit only the necessary parts for enabling that. Let's try to be focused on that, and leave the original code-base refactoring for later phases.

DaniBodor commented:
> Great, now running the tests will be much lighter :D
> 
> > * some test functions were basically used as the callback function of a "test all envs" function. I made those internal only functions, to avoid return statement warnings (this would become errors in the next pytest update).
> 
> Why they would become errors in the next pytest update?

The warning message said they would :)

> > * previous test_speed was not an actual test, but more of a way to log how long things take. Also, it was quite slow. So I renamed it, thereby avoiding it being run each time pytest is run.
> 
> As it is now, we won't notice any error when changing the code, and we will need to remember to run it every time to keep it up-to-date with the code-base. I'd either remove it or leave it as a test (naming it back to test_speed).

The point is that it is not a test (in terms of checking for errors), but a script that reports how long certain procedures take. So the whole point of renaming it is to not run it each time we run pytest. But still keep it to allow for reporting on performance if needed. It seems useful to keep. Maybe it should go into a different folder?

> > NOT SOLVED: A number of the tests (especially in `test_env`) catch all errors in the middle of the test function. This means that the test cannot fail and is basically not testing anything (instead in `print`s an error message, but pytest doesn't show that). I tried not catching errors, but then the test does fail and I didn't look into it any further (yet). I added FIXME statements to those.
> 
> Useful that you have already flagged it. In general, I'd avoid looking too much into such issues for now, since the priority at the moment is to let the package work for our experiments, and edit only the necessary parts for enabling that. Let's try to be focused on that, and leave the original code-base refactoring for later phases.

Indeed, which is why I only made the minimal effort of testing what happens if the errors are not suppressed, but that indeed led to failing tests.

Commit:
test: consolidate warnings catches
